mlir/lib/Bindings/Python/IRAttributes.cpp

```diff
@@ -1016,14 +930,177 @@ class PyDenseElementsAttribute
+      } else if (isSignedIntegerFormat(format)) {
+        if (view.itemsize == 4) {
+          // i32
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 32)
+                                    : mlirIntegerTypeSignedGet(context, 32);
+        } else if (view.itemsize == 8) {
+          // i64
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 64)
+                                    : mlirIntegerTypeSignedGet(context, 64);
+        } else if (view.itemsize == 1) {
+          // i8
+          bulkLoadElementType = signless ? mlirIntegerTypeGet(context, 8)
+                                         : mlirIntegerTypeSignedGet(context, 8);
+        } else if (view.itemsize == 2) {
+          // i16
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 16)
+                                    : mlirIntegerTypeSignedGet(context, 16);
+        }
+      } else if (isUnsignedIntegerFormat(format)) {
+        if (view.itemsize == 4) {
+          // unsigned i32
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 32)
+                                    : mlirIntegerTypeUnsignedGet(context, 32);
+        } else if (view.itemsize == 8) {
+          // unsigned i64
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 64)
+                                    : mlirIntegerTypeUnsignedGet(context, 64);
+        } else if (view.itemsize == 1) {
+          // i8
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 8)
+                                    : mlirIntegerTypeUnsignedGet(context, 8);
+        } else if (view.itemsize == 2) {
+          // i16
+          bulkLoadElementType = signless
+                                    ? mlirIntegerTypeGet(context, 16)
+                                    : mlirIntegerTypeUnsignedGet(context, 16);
+        }
+      }
+      if (!bulkLoadElementType) {
+        throw std::invalid_argument(
+            std::string("unimplemented array format conversion from format: ") +
+            std::string(format));
+      }
+    }
+
+    MlirType type = getShapedType(bulkLoadElementType, explicitShape, view);
+    return mlirDenseElementsAttrRawBufferGet(type, view.len, view.buf);
+  }
+
+  // There is a complication for boolean numpy arrays, as numpy represent them
+  // as 8 bits per boolean, whereas MLIR bitpacks them into 8 booleans per byte.
+  // This function does the bit-packing respecting endianess.
+  static MlirAttribute getBitpackedAttributeFromBooleanBuffer(
+      Py_buffer &view, std::optional<std::vector<int64_t>> explicitShape,
+      MlirContext &context) {
+    // First read the content of the python buffer as u8's, to correct for
+    // endianess
```

Comment:
sorry don't understand? what/why does this need to be "corrected"? I can see you're setting `little` below but shouldn't you be preserving/using the endianness for the host arch (wherever the bindings are running...)? probably I don't understand something here...

Reply:
From https://mlir.llvm.org/docs/BytecodeFormat/#fixed-width-integers my understanding is that MLIR attributes always have their data stored in a little-endian format. If that is not true for MlirAttributes in general, the code is wrong.

Reply:
@stellaraccident this is the one i'm most uncertain about

Reply:
That's for bytecode rather than in memory. If this is converting the in memory format, then that section doesn't apply.

Reply:
Ack, one thing I don't understand though, is that if I follow the `mlirDenseElementsAttrRawBufferGet` function, it goes to `DenseElementsAttr::getFromRawBuffer`, and I believe it eventually ends up in `BuiltinAttributes.cpp` where it uses the [`writeBits` and `readBits` functions](https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/BuiltinAttributes.cpp#L541) respectively. Both of these have special handling for big endian, where it seems to convert the format.

It is possible that it ends up [calling `Base::get` instead](https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/BuiltinAttributes.cpp#L1354), which may be why the above is not true. My C++ navigation skills is currently failing me, to follow which code this `Base::get` call will execute.

Reply:
Do you know of a way to test this? From reading around other llvm docs, it seems like user-mode qemu is the common approach for testing big-endian systems. Do any of you have experience with that and the Python API?

Reply:
> it goes to `DenseElementsAttr::getFromRawBuffer`

https://github.com/makslevental/llvm-project/blob/84aa02d3fa1f1f614c4f3c144ec118b2f05ae6b0/mlir/lib/IR/BuiltinAttributes.cpp#L1064

which does indeed hit `Base::get`

https://github.com/makslevental/llvm-project/blob/84aa02d3fa1f1f614c4f3c144ec118b2f05ae6b0/mlir/lib/IR/BuiltinAttributes.cpp#L1354

which then goes to `static ConcreteT StorageUserBase::get`

https://github.com/makslevental/llvm-project/blob/a9636b7f60f283926c66e96c036f5b5d9e57c026/mlir/include/mlir/IR/StorageUniquerSupport.h#L177

I think possibly you could just remove the explicit use of `little` in the various places and defer to host endiannes? I mean really this is all academic because as I understand it, there are no extant big-endian systems we care about? Someone with more experience can comment on this hypothesis...

Reply:
Thanks for the pointers!

Unfortunately, it will not work simply remove the "little" part from the numpy calls, as it defaults to big-endian and not the native byteorder.

As you mention, I doubt any big-endian users will ever call this code anyways. What do you think about just throwing a "this is unsupported" exception (which is the current behaviour) if someone calls it on a big-endian system?

I think that check can be made something like:
```c++
if (llvm::endianness::native == llvm::endianness::big) {
    throw py::type_error("Boolean types are unsupported on big-endian systems");
}
```

I could also use the above check to switch between "little"/"big" endian in the numpy call, but I am considering if throwing is actually better, given we have no way to really test it?

Reply:
Ya throwing here is acceptable and reasonable (if why IBM customers ever show up we can discuss it then 😅)